Fix off-by-one error in String.indexOf acceleration on z15

fjeremic opened the pull request:
- `stringIndexReg` contains the current index into the search string
  that we will start the search on
- `maxIndexReg = searchString.length() - patternString.length()`
  and this value does not change at all
- `loadLenReg` is the minimum of the number of characters left to load
  from the search string or 16. That is `loadLenReg` <= 16.

The problem here is that due to the comparison we are making we break
the assumption that the remaining number of characters we have left to
examine is always greater than or equal to the length of the pattern.
The code snippet being changed has two branches:

- The first branch branches if `stringIndexReg >= maxIndexReg`
- The second branch branches if `stringIndexReg >= searchString.length()`

There are two problems with the first branch. First problem is that
`stringIndexReg` is updated after the branch, when it should be updated
before the branch. Second issue is that we can have a false negative
result because we branch on equality when we really shouldn't.

The second branch is actually not needed since it will never be executed
because `maxIndexReg` is always <= `searchString.length()`. This branch
is removed.

There is another issue in case there is a full pattern match. In the
full pattern match case, the first character of the match will be at
index 0, but the path following `labelPatternHeadFullMatch` in the code
assumes we have enough characters left in the search string to search
for the entire pattern. This may not be the case. To cover this scenario
we add another check in this path to ensure we have enough characters
left to carry out the full pattern search.

The above problem with the full pattern match does not exist for the
partial pattern match because the partial pattern match handling code
updates the search index so that we begin the match at index 0 and 
we re-run the matching (VSTRS) code again. After re-running we will
either get a full match or no match. Both of these cases have been
fixed up as per above notes. Thus we are fully covered.

fjeremic commented:
Jenkins test sanity zlinux jdk8,jdk11

fjeremic commented:
Jenkins test sanity zlinux jdk8,jdk11

fjeremic commented on runtime/compiler/z/codegen/J9TreeEvaluator.cpp:
```diff
@@ -741,6 +741,7 @@ J9::Z::TreeEvaluator::inlineVectorizedStringIndexOf(TR::Node* node, TR::CodeGene
       cursor = generateRIInstruction(cg, TR::InstOpCode::getAddHalfWordImmOpCode(), node, loadLenReg, -1);
       iComment("needs -1 because VLL's third operand is the highest index to load");
       generateVRSbInstruction(cg, TR::InstOpCode::VLL, node, stringVReg, loadLenReg, generateS390MemoryReference(stringCharPtrReg, headerSize, cg));
+      generateRIInstruction(cg, TR::InstOpCode::getAddHalfWordImmOpCode(), node, loadLenReg, 1);
```
The reason this change is needed is because we have two paths from `labelFindPatternHead`.

1. There are < 16 characters left to load from search string
2. There are >= 16 characters left to load from search string

In (2) the value of `loadLenReg` is always 16, and is the number of characters we loaded. In (1) however the `loadLenReg` register contains one less than the number of characters we ended up loading because of the way `VLL` instruction works. We must compensate for this by incrementing this register after the load so that the logic after `VSTRS` is consistent in that  `loadLenReg` contains the number of characters we just loaded.

joransiu replied:
Agreed with the need to add 1 to re-adjust `loadLenReg` for case (1).  Thanks for the added explanation.

fjeremic commented:
Jenkins test sanity zlinux jdk11

fjeremic commented:
Noting that all JitHelpers tests passed which contain unit tests for this intrinsic as can be seen here:

https://github.com/eclipse-openj9/openj9/blob/5d2495e99a7811e81fb3f04a1088791db0db2652/test/functional/Java8andUp/src/org/openj9/test/com/ibm/jit/Test_JITHelpers.java#L436-L1148

Links to the jobs:
- https://openj9-jenkins.osuosl.org/job/Test_openjdk11_j9_sanity.functional_s390x_linux_Personal_testList_1/84/consoleText
- https://openj9-jenkins.osuosl.org/job/Test_openjdk11_j9_sanity.functional_s390x_linux_Personal_testList_0/84/consoleText
- https://openj9-jenkins.osuosl.org/job/Test_openjdk8_j9_sanity.functional_s390x_linux_Personal_testList_0/76/consoleText
- https://openj9-jenkins.osuosl.org/job/Test_openjdk8_j9_sanity.functional_s390x_linux_Personal_testList_1/76/consoleText

Scraped job output:
```
[2021-08-22T02:37:09.886Z] ===============================================
[2021-08-22T02:37:09.886Z] Running test JCL_Test_JITHelpers_1 ...
[2021-08-22T02:37:09.886Z] ===============================================
...
[2021-08-22T02:37:29.965Z] PASSED: test_getSuperclass
[2021-08-22T02:37:29.965Z] PASSED: test_intrinsicIndexOfLatin1
[2021-08-22T02:37:29.965Z] PASSED: test_intrinsicIndexOfStringLatin1
[2021-08-22T02:37:29.965Z] PASSED: test_intrinsicIndexOfStringUTF16
[2021-08-22T02:37:29.965Z] PASSED: test_intrinsicIndexOfUTF16
[2021-08-22T02:37:29.965Z] PASSED: test_toLowerIntrinsicLatin1
[2021-08-22T02:37:29.965Z] PASSED: test_toLowerIntrinsicLatin1_edgecaseLengths
[2021-08-22T02:37:29.965Z] PASSED: test_toLowerIntrinsicUTF16
[2021-08-22T02:37:29.965Z] PASSED: test_toLowerIntrinsicUTF16_edgecaseLengths
[2021-08-22T02:37:29.965Z] PASSED: test_toUpperIntrinsicLatin1
[2021-08-22T02:37:29.965Z] PASSED: test_toUpperIntrinsicLatin1_edgecaseLengths
[2021-08-22T02:37:29.965Z] PASSED: test_toUpperIntrinsicUTF16
[2021-08-22T02:37:29.965Z] PASSED: test_toUpperIntrinsicUTF16_edgecaseLengths
[2021-08-22T02:37:29.965Z] 
[2021-08-22T02:37:29.965Z] ===============================================
[2021-08-22T02:37:29.965Z]     JCL_TEST_JITHelpers
[2021-08-22T02:37:29.965Z]     Tests run: 13, Failures: 0, Skips: 0
[2021-08-22T02:37:29.965Z] ===============================================
...
[2021-08-22T02:37:31.632Z] ===============================================
[2021-08-22T02:37:31.632Z] Running test JCL_Test_JITHelpers_2 ...
[2021-08-22T02:37:31.632Z] ===============================================
[2021-08-22T02:37:46.294Z] PASSED: test_getSuperclass
[2021-08-22T02:37:46.294Z] PASSED: test_intrinsicIndexOfLatin1
[2021-08-22T02:37:46.294Z] PASSED: test_intrinsicIndexOfStringLatin1
[2021-08-22T02:37:46.294Z] PASSED: test_intrinsicIndexOfStringUTF16
[2021-08-22T02:37:46.294Z] PASSED: test_intrinsicIndexOfUTF16
[2021-08-22T02:37:46.294Z] PASSED: test_toLowerIntrinsicLatin1
[2021-08-22T02:37:46.294Z] PASSED: test_toLowerIntrinsicLatin1_edgecaseLengths
[2021-08-22T02:37:46.294Z] PASSED: test_toLowerIntrinsicUTF16
[2021-08-22T02:37:46.294Z] PASSED: test_toLowerIntrinsicUTF16_edgecaseLengths
[2021-08-22T02:37:46.294Z] PASSED: test_toUpperIntrinsicLatin1
[2021-08-22T02:37:46.294Z] PASSED: test_toUpperIntrinsicLatin1_edgecaseLengths
[2021-08-22T02:37:46.294Z] PASSED: test_toUpperIntrinsicUTF16
[2021-08-22T02:37:46.294Z] PASSED: test_toUpperIntrinsicUTF16_edgecaseLengths
[2021-08-22T02:37:46.294Z] 
[2021-08-22T02:37:46.294Z] ===============================================
[2021-08-22T02:37:46.294Z]     JCL_TEST_JITHelpers
[2021-08-22T02:37:46.295Z]     Tests run: 13, Failures: 0, Skips: 0
[2021-08-22T02:37:46.295Z] ===============================================
...
[2021-08-20T19:49:35.689Z] ===============================================
[2021-08-20T19:49:35.689Z] Running test JCL_Test_JITHelpers_SE80_2 ...
[2021-08-20T19:49:35.689Z] ===============================================
...
[2021-08-20T19:49:46.569Z] PASSED: test_getSuperclass
[2021-08-20T19:49:46.569Z] PASSED: test_intrinsicIndexOfLatin1
[2021-08-20T19:49:46.569Z] PASSED: test_intrinsicIndexOfStringLatin1
[2021-08-20T19:49:46.569Z] PASSED: test_intrinsicIndexOfStringUTF16
[2021-08-20T19:49:46.569Z] PASSED: test_intrinsicIndexOfUTF16
[2021-08-20T19:49:46.569Z] PASSED: test_toLowerIntrinsicLatin1
[2021-08-20T19:49:46.569Z] PASSED: test_toLowerIntrinsicLatin1_edgecaseLengths
[2021-08-20T19:49:46.569Z] PASSED: test_toLowerIntrinsicUTF16
[2021-08-20T19:49:46.569Z] PASSED: test_toLowerIntrinsicUTF16_edgecaseLengths
[2021-08-20T19:49:46.569Z] PASSED: test_toUpperIntrinsicLatin1
[2021-08-20T19:49:46.569Z] PASSED: test_toUpperIntrinsicLatin1_edgecaseLengths
[2021-08-20T19:49:46.569Z] PASSED: test_toUpperIntrinsicUTF16
[2021-08-20T19:49:46.569Z] PASSED: test_toUpperIntrinsicUTF16_edgecaseLengths
[2021-08-20T19:49:46.569Z] 
[2021-08-20T19:49:46.569Z] ===============================================
[2021-08-20T19:49:46.569Z]     JCL_TEST_JITHelpers
[2021-08-20T19:49:46.569Z]     Tests run: 13, Failures: 0, Skips: 0
[2021-08-20T19:49:46.569Z] ===============================================
...
[2021-08-20T19:50:03.915Z] ===============================================
[2021-08-20T19:50:03.915Z] Running test JCL_Test_JITHelpers_SE80_0 ...
[2021-08-20T19:50:03.915Z] ===============================================
...
[2021-08-20T19:50:16.637Z] PASSED: test_getSuperclass
[2021-08-20T19:50:16.637Z] PASSED: test_intrinsicIndexOfLatin1
[2021-08-20T19:50:16.637Z] PASSED: test_intrinsicIndexOfStringLatin1
[2021-08-20T19:50:16.637Z] PASSED: test_intrinsicIndexOfStringUTF16
[2021-08-20T19:50:16.637Z] PASSED: test_intrinsicIndexOfUTF16
[2021-08-20T19:50:16.637Z] PASSED: test_toLowerIntrinsicLatin1
[2021-08-20T19:50:16.637Z] PASSED: test_toLowerIntrinsicLatin1_edgecaseLengths
[2021-08-20T19:50:16.637Z] PASSED: test_toLowerIntrinsicUTF16
[2021-08-20T19:50:16.637Z] PASSED: test_toLowerIntrinsicUTF16_edgecaseLengths
[2021-08-20T19:50:16.637Z] PASSED: test_toUpperIntrinsicLatin1
[2021-08-20T19:50:16.637Z] PASSED: test_toUpperIntrinsicLatin1_edgecaseLengths
[2021-08-20T19:50:16.637Z] PASSED: test_toUpperIntrinsicUTF16
[2021-08-20T19:50:16.637Z] PASSED: test_toUpperIntrinsicUTF16_edgecaseLengths
[2021-08-20T19:50:16.637Z] 
[2021-08-20T19:50:16.637Z] ===============================================
[2021-08-20T19:50:16.638Z]     JCL_TEST_JITHelpers
[2021-08-20T19:50:16.638Z]     Tests run: 13, Failures: 0, Skips: 0
[2021-08-20T19:50:16.638Z] ===============================================
```

fjeremic commented:
Next I will add unit tests to capture the original bug and back out the fix and re-run to tests to show explicitly that the unit tests will fail.

fjeremic commented:
Launching the tests now. Note we expect to see failures here since the fix has been reverted temporarily.

Jenkins test sanity zlinux jdk8,jdk11

fjeremic commented:
> Launching the tests now. Note we expect to see failures here since the fix has been reverted temporarily.

Thinking of this again, we may not get failures unless we run on z15. Let's wait and see if we get lucky, otherwise I'll have to launch a grinder on z15 explicitly.

fjeremic commented:
It seems we did end up running on z15 and we do have failures as observed:

- https://openj9-jenkins.osuosl.org/job/Test_openjdk11_j9_sanity.functional_s390x_linux_Personal_testList_0/85/consoleText
```
[2021-08-23T19:28:25.096Z] FAILED: test_intrinsicIndexOfStringUTF16
[2021-08-23T19:28:25.096Z] java.lang.AssertionError: expected [-1] but found [20]
[2021-08-23T19:28:25.096Z] 	at org.testng.Assert.fail(Assert.java:96)
[2021-08-23T19:28:25.096Z] 	at org.testng.Assert.failNotEquals(Assert.java:776)
[2021-08-23T19:28:25.096Z] 	at org.testng.Assert.assertEqualsImpl(Assert.java:137)
[2021-08-23T19:28:25.096Z] 	at org.testng.Assert.assertEquals(Assert.java:118)
[2021-08-23T19:28:25.096Z] 	at org.testng.Assert.assertEquals(Assert.java:652)
[2021-08-23T19:28:25.096Z] 	at org.testng.Assert.assertEquals(Assert.java:662)
[2021-08-23T19:28:25.096Z] 	at org.openj9.test.com.ibm.jit.Test_JITHelpers.test_intrinsicIndexOfStringUTF16(Test_JITHelpers.java:1146)
```

- https://openj9-jenkins.osuosl.org/job/Test_openjdk11_j9_sanity.functional_s390x_linux_Personal_testList_1/85/consoleText
```
[2021-08-23T19:28:11.273Z] java.lang.AssertionError: expected [-1] but found [21]
[2021-08-23T19:28:11.273Z] 	at org.testng.Assert.fail(Assert.java:96)
[2021-08-23T19:28:11.273Z] 	at org.testng.Assert.failNotEquals(Assert.java:776)
[2021-08-23T19:28:11.273Z] 	at org.testng.Assert.assertEqualsImpl(Assert.java:137)
[2021-08-23T19:28:11.273Z] 	at org.testng.Assert.assertEquals(Assert.java:118)
[2021-08-23T19:28:11.273Z] 	at org.testng.Assert.assertEquals(Assert.java:652)
[2021-08-23T19:28:11.273Z] 	at org.testng.Assert.assertEquals(Assert.java:662)
[2021-08-23T19:28:11.273Z] 	at org.openj9.test.com.ibm.jit.Test_JITHelpers.test_intrinsicIndexOfStringLatin1(Test_JITHelpers.java:907)
```

- https://openj9-jenkins.osuosl.org/job/Test_openjdk8_j9_sanity.functional_s390x_linux_Personal_testList_1/77/consoleText
```
[2021-08-23T19:16:27.664Z] FAILED: test_intrinsicIndexOfStringUTF16
[2021-08-23T19:16:27.664Z] java.lang.AssertionError: expected [-1] but found [20]
[2021-08-23T19:16:27.664Z] 	at org.testng.Assert.fail(Assert.java:96)
[2021-08-23T19:16:27.664Z] 	at org.testng.Assert.failNotEquals(Assert.java:776)
[2021-08-23T19:16:27.664Z] 	at org.testng.Assert.assertEqualsImpl(Assert.java:137)
[2021-08-23T19:16:27.664Z] 	at org.testng.Assert.assertEquals(Assert.java:118)
[2021-08-23T19:16:27.664Z] 	at org.testng.Assert.assertEquals(Assert.java:652)
[2021-08-23T19:16:27.664Z] 	at org.testng.Assert.assertEquals(Assert.java:662)
[2021-08-23T19:16:27.664Z] 	at org.openj9.test.com.ibm.jit.Test_JITHelpers.test_intrinsicIndexOfStringUTF16(Test_JITHelpers.java:1146)
```

Now I'm going to back out the revert fix (1d9a9de) and launch the tests again. We should expect all of these tests to pass then.

fjeremic commented:
Jenkins test sanity zlinux jdk8,jdk11

fjeremic commented:
Tests pass as expected. Removing WIP as this is ready for review. @joransiu FYI.

joransiu commented on runtime/compiler/z/codegen/J9TreeEvaluator.cpp:
```diff
@@ -764,9 +765,8 @@ J9::Z::TreeEvaluator::inlineVectorizedStringIndexOf(TR::Node* node, TR::CodeGene
-      generateRIEInstruction(cg, TR::InstOpCode::getCmpRegAndBranchRelOpCode(), node, stringIndexReg, maxIndexReg, labelPatternNotFound, TR::InstOpCode::COND_BNL);
       generateRRInstruction(cg, TR::InstOpCode::getAddRegOpCode(), node, stringIndexReg, loadLenReg);
-      generateRIEInstruction(cg, TR::InstOpCode::getCmpRegAndBranchRelOpCode(), node, stringIndexReg, stringLenReg, labelPatternNotFound, TR::InstOpCode::COND_BNL);
+      generateRIEInstruction(cg, TR::InstOpCode::getCmpRegAndBranchRelOpCode(), node, stringIndexReg, maxIndexReg, labelPatternNotFound, TR::InstOpCode::COND_BH);
```
Might be worth adding an iComment to explain this comparison.  i.e. "Updated stringIndex for next iteration exceeds maxIndex of valid match.  Full pattern cannot be matched."

fjeremic replied:
Fixed in [force-push](https://github.com/eclipse-openj9/openj9/compare/b2389b0722853a5cac394cbb8948c6e8ebdc9cdc..c20200d).

joransiu commented on runtime/compiler/z/codegen/J9TreeEvaluator.cpp:
```diff
@@ -781,6 +781,7 @@ J9::Z::TreeEvaluator::inlineVectorizedStringIndexOf(TR::Node* node, TR::CodeGene
       generateRRInstruction(cg, TR::InstOpCode::getAddRegOpCode(), node, matchIndexReg, stringIndexReg);
       cursor = generateRIEInstruction(cg, TR::InstOpCode::getCmpImmBranchRelOpCode(), node, patternLenReg, (int8_t)vectorSize, labelLoadResult, TR::InstOpCode::COND_BNH);
       iComment("if patternLen <= 16 then we are done, otherwise we continue to check the rest of pattern");
+      generateRIEInstruction(cg, TR::InstOpCode::getCmpRegAndBranchRelOpCode(), node, stringIndexReg, maxIndexReg, labelPatternNotFound, TR::InstOpCode::COND_BH);
```
Similarly, an iComment maybe useful here. i.e. something like: "Updated stringIndex for start of matched section exceeds maxIndex. Full pattern cannot be matched."

fjeremic replied:
Fixed in [force-push](https://github.com/eclipse-openj9/openj9/compare/b2389b0722853a5cac394cbb8948c6e8ebdc9cdc..c20200d).

joransiu reviewed (approved):
Code changes LGTM.  Thanks for the added tests.  

One thing about the commit message:
> - `loadLenReg` at this point in code is always 16`

I don't think that's fully correct.  It really should be as you described in the comment on line 744.  If < 16 chars left to load from search string, it'll be the remaining length.  IF >= 16, loadLengReg == 16.

fjeremic commented:
@joransiu I've addressed the review comments and updated the commit message to be much more descriptive. Launching the full tests on your behalf.

Jenkins test sanity all jdk8,jdk11

joransiu reviewed (approved):
Latest changes and updated commit message all LGTM!   Thanks @fjeremic
